lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -851,21 +851,39 @@ impl Drop for CGOPNG {
 }
 
 #[cfg(unix)]
-fn wait_for_fd(fd: usize) -> bool {
-    unsafe {
-        let mut raw_fds: fd_set = mem::zeroed();
-
-        FD_SET(fd as i32, &mut raw_fds);
-
-        let result = select(
-            fd as i32 + 1,
-            &mut raw_fds,
-            ptr::null_mut(),
-            ptr::null_mut(),
-            ptr::null_mut(),
-        );
-        result == 1
+fn wait_for_fd(fd: usize) -> RdpResult<()> {
+    let fds = &mut libc::pollfd {
+        fd: fd as i32,
+        events: libc::POLLIN,
+        revents: 0,
+    };
+    let res = unsafe { libc::poll(fds, 1, -1) };
+
+    // We only use one fd and we should never timeout.
+    if res == 1 {
+        // POLLIN means that the fd is ready to be read from,
+        // POLLHUP means that the other side of the pipe was closed,
+        // but we still may have data to read.
+        if fds.revents & (libc::POLLIN | libc::POLLHUP) != 0 {
+            return Ok(());
+        } else if fds.revents & libc::POLLNVAL != 0 {
+            return Err(RdpError::Io(IoError::new(
+                std::io::ErrorKind::InvalidInput,
+                "invalid fd",
+            )));
+        } else if fds.revents & libc::POLLERR != 0 {
+            return Err(RdpError::Io(IoError::new(
+                std::io::ErrorKind::Other,
+                "error on fd",
+            )));
+        }
     }
+
+    // res == -1
+    Err(RdpError::Io(IoError::new(
+        std::io::ErrorKind::Other,
+        "poll failed",
+    )))
 }
```

Comment:
Very slight nit, it might be nice to include `res` in the message on the off chance some future refactoring results in some `poll()` result other than -1/1.

Reply:
Added it to a log 3f1fedddfeace5eb96dbcdbe0893321b79221ae0